### PR TITLE
drivers:platform:aducm3029:gpio Add GPIO support

### PR DESCRIPTION
**Pull request by mchindri**
Add GPIO support for ADuCM3029

Signed-off-by: Mihail Chindris <mihail.chindris@analog.com>

### CODE REVIEW

File: drivers/platform/aducm3029/gpio.c [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+		return FAILURE;
+
+	(*desc) = calloc(1, sizeof(**desc));
+	if (!(*desc))
+		return FAILURE;
+
+	(*desc)->number = param->number;
+
+	/* If this is the first GPIO initialize GPIO controller */
+	if(nb_gpio == 0)
```
Comment (cristiantpop) [inline drivers/platform/aducm3029/gpio.c]: Add space after if, before bracket.

File: drivers/platform/aducm3029/gpio.c [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+ */
+int32_t gpio_direction_input(struct gpio_desc *desc)
+{
+	if (!desc || !nb_gpio)
+		return FAILURE;
+	/* Enable input driver */
+	if (ADI_GPIO_SUCCESS != adi_gpio_InputEnable(PORT(desc->number),
+			PIN(desc->number), true))
+		return FAILURE;
+	return SUCCESS;
```
Comment (cristiantpop) [inline drivers/platform/aducm3029/gpio.c]: Add new line before return.

File: drivers/platform/aducm3029/gpio.c [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+		return FAILURE;
+
+	/* Enable output driver */
+	ret = adi_gpio_OutputEnable(PORT(desc->number), PIN(desc->number),
+				    true);
+	if(ret != ADI_GPIO_SUCCESS)
+		return FAILURE;
+
+	/* Initialize pin with a value */
+	if(value == 1)
```
Comment (cristiantpop) [inline drivers/platform/aducm3029/gpio.c]: Add space after if, before bracket. Check in all locations.

File: drivers/platform/aducm3029/gpio.c [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+	ADI_GPIO_RESULT ret;
+
+	if (!desc || !nb_gpio)
+		return FAILURE;
+
+	if(value == GPIO_HIGH)
+		ret = adi_gpio_SetHigh(PORT(desc->number), PIN(desc->number));
+	else
+		ret = adi_gpio_SetLow(PORT(desc->number), PIN(desc->number));
+	if (ret != ADI_GPIO_SUCCESS)
```
Comment (amiclaus) [inline drivers/platform/aducm3029/gpio.c]: nitpick: add space before `if` condition

File: drivers/platform/aducm3029/gpio.c [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+		return FAILURE;
+	return SUCCESS;
```
Comment (amiclaus) [inline drivers/platform/aducm3029/gpio.c]: nitpick: add space before `return SUCCESS` condition

File: drivers/platform/aducm3029/gpio.c [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+		return FAILURE;
+
+	if (ADI_GPIO_SUCCESS != adi_gpio_GetOutputEnable(PORT(desc->number),
+			&pins))
+		return FAILURE;
+	if (pins & PIN(desc->number))
+		*direction = GPIO_OUT;
+	else
+		*direction = GPIO_IN;
+	return SUCCESS;
```
Comment (amiclaus) [inline drivers/platform/aducm3029/gpio.c]: nitpick: add empty line before `return SUCCESS`

File: drivers/platform/aducm3029/gpio.c [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+		return FAILURE;
+
+	/* Initialize pin with a value */
+	if(value == 1)
+		ret = adi_gpio_SetHigh(PORT(desc->number), PIN(desc->number));
+	else
+		ret = adi_gpio_SetLow(PORT(desc->number), PIN(desc->number));
+	if(ret != ADI_GPIO_SUCCESS)
+		return FAILURE;
+	return SUCCESS;
```
Comment (amiclaus) [inline drivers/platform/aducm3029/gpio.c]: nitpick: add empty line before `return SUCCESS`

File: drivers/platform/aducm3029/gpio.c [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+		return FAILURE;
+
+	if (ADI_GPIO_SUCCESS != adi_gpio_GetData(PORT(desc->number),
+			PIN(desc->number), &pins))
+		return FAILURE;
+	if (pins & PIN(desc->number))
+		*value = GPIO_HIGH;
+	else
+		*value = GPIO_LOW;
+	return SUCCESS;
```
Comment (amiclaus) [inline drivers/platform/aducm3029/gpio.c]: nitpick: add empty line before `return SUCCESS`

File: drivers/platform/aducm3029/gpio.c [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+ * @return \ref SUCCESS in case of success, \ref FAILURE otherwise.
+ */
+int32_t gpio_set_value(struct gpio_desc *desc, uint8_t value)
+{
+	ADI_GPIO_RESULT ret;
+
+	if (!desc || !nb_gpio)
+		return FAILURE;
+
+	if(value == GPIO_HIGH)
```
Comment (adrimbarean) [inline drivers/platform/aducm3029/gpio.c]: nitpick: might be a little more intuitive if the `if` clause contained the `GPIO_LOW` situation and the `else` clause the rest. As it stands if value is 1 then te GPIO goes high and if value is 0, 2, 3, etc. will go low.

File: drivers/platform/aducm3029/gpio.c [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+	if (!desc || !nb_gpio)
+		return FAILURE;
+
+	if (ADI_GPIO_SUCCESS != adi_gpio_GetData(PORT(desc->number),
+			PIN(desc->number), &pins))
+		return FAILURE;
+	if (pins & PIN(desc->number))
+		*value = GPIO_HIGH;
+	else
+		*value = GPIO_LOW;
```
Comment (adrimbarean) [inline drivers/platform/aducm3029/gpio.c]: As it stands `adi_gpio_GetData` can only be accessed with one pin. This means that it should be OK to write:
`*value = pins;`
or to be safer and still compress the code:
`*value = !!pins;`

File: drivers/platform/aducm3029/gpio.c (hunk 1 above)
Comment (mchindri) [reply drivers/platform/aducm3029/gpio.c]: Less spaces- > less memory for the source is needed.
Joke, solved :)

File: drivers/platform/aducm3029/gpio.c (hunk 9 above)
Comment (mchindri) [reply drivers/platform/aducm3029/gpio.c]: changed to `if (value == GPIO_LOW)`
is better this way?

File: drivers/platform/aducm3029/gpio.c (hunk 10 above)
Comment (mchindri) [reply drivers/platform/aducm3029/gpio.c]: true

**Comment by mchindri** [comment]
V2
- add spaces where needed
- remove empty lines
- change check in gpio_set_value to check if GPIO_LOW
- imporove gpio_get_value-> *value = !!pins;

**Pull request merged**